rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+        }
+    }
+
+    pub mod block_region_table {
+        // invariant: all blocks whose IDs are below the total_allocated_blocks are valid.
+
+        use super::{offset, size};
+        use crate::ic0_stable::nicer::{read_u16, write_u16};
+        use crate::region::{BlockId, RegionId};
+
+        // Compute an offset in stable memory for a particular block ID (based zero).
+        fn index(b: &BlockId) -> u64 {
+            offset::BLOCK_REGION_TABLE + (b.0 as u64 * size::BLOCK_REGION_TABLE_ENTRY as u64)
+        }
+
+        /// Some(r, j) means that the block is in use by region r, at slot j.
+        /// None means that the block is available for (re-)allocation.
+        pub fn get(b: BlockId) -> Option<(RegionId, u16)> {
+            let raw = read_u16(index(&b));
+            let rid = RegionId::from_u16(raw);
+            rid.map(|r| (r, read_u16(index(&b) + 2)))
+        }
+
+        /// Some(r, j) means that the block is in use by region r, at slot j.
+        /// None means that the block is available for (re-)allocation.
+        pub fn set(b: BlockId, r: Option<(RegionId, u16)>) {
+            match r {
+                None => write_u16(index(&b), RegionId::into_u16(None)),
+                Some((r, j)) => {
+                    write_u16(index(&b), RegionId::into_u16(Some(r)));
+                    write_u16(index(&b) + 2, j)
+                }
+            }
+        }
+    }
+
+    pub mod region_table {
+        use crate::region::{RegionId, RegionSizeInPages};
+
+        // invariant (for now, pre-GC integration):
+        //  all regions whose IDs are below the total_allocated_regions are valid.
+        use super::{offset, size};
+        use crate::ic0_stable::nicer::{read_u64, write_u64};
+
+        fn index(r: &RegionId) -> u64 {
+            offset::REGION_TABLE + r.0 as u64 * size::REGION_TABLE_ENTRY as u64
+        }
+
+        /// Some(_) gives the size in pages.
+        /// None means that the region is not in use.
+        pub fn get(r: &RegionId) -> Option<RegionSizeInPages> {
+            RegionSizeInPages::from_u64(read_u64(index(r)))
+        }
+
+        pub fn set(r: &RegionId, s: Option<RegionSizeInPages>) {
+            write_u64(index(r), RegionSizeInPages::into_u64(s))
+        }
+    }
+}
+
+unsafe fn alloc_region<M: Memory>(
+    mem: &mut M,
+    id: u16,
+    page_count: u32,
+    vec_pages: Value,
+) -> Value {
+    let r_ptr = mem.alloc_words(size_of::<Region>());
+
+    // NB. cannot use as_region() here as we didn't write the header yet
+    let region = r_ptr.get_ptr() as *mut Region;
+    (*region).header.tag = TAG_REGION;
+    (*region).header.init_forward(r_ptr);
+    (*region).id = id;
+    // The padding must be initialized with zero because it is read by the compiler-generated code.
+    (*region).zero_padding = 0;
+    (*region).page_count = page_count;
+    init_with_barrier(mem, &mut (*region).vec_pages, vec_pages);
+
+    r_ptr
```

Comment:
Needed for correctness of the incremental GC:

```suggestion
    allocation_barrier(r_ptr);
    r_ptr
```